<a href="https://www.kaggle.com/code/ankumagawa/distance-metrics-cosine-manhattan-eucclidean?scriptVersionId=92892566" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/medical-ratio1/Medical Ratio.csv


Distance metrics are a key part of several machine learning algorithms. A distance measure is an objective score that summarizes the relative difference between two objects in a problem domain.


These distance metrics are used in both supervised and unsupervised learning, generally to calculate the similarity between data points.

Let’s say we want to create clusters using the K-Means Clustering or k-Nearest Neighbour algorithm to solve a classification or regression problem. How will you define the similarity between different observations here? How can we say that two points are similar to each other?

We can calculate the distance between points and then define the similarity between them. How do we calculate this distance and what are the different distance metrics in machine learning?


There are several ways of measuring distance between point a and b in d dimensions with closer distance implying similarity.

In this project, I will introduce different approaches to measuring distance. These approachses are:

1) Minkowski Distance Metric
    
    * Mahanttan Distance:
    
    * Euclidean Distance:

2) Weighted Similarity

3) Cosine Similarity

4) Kullback-Leibler Divergence

5) Jensen-Shannon

Follow-up section is the code implementation with the case study is 37 pharmarceutical companies and theirs number of patent applied in each IPC. The goal of this section is how to apply different Similarity Measure Algorithm to calculate the Technonoly Similarity between the companies. 

# Kernel Layout:

# <a href='#I'>  1) Thereotical and Mathematic Background: </a>

## <a href='#1.1'> 1.1) Minkowski  Distance Metric: </a>

## <a href='#1.2'> 1.2) Eucclidean Distance Metric: </a>

## <a href='#1.3'> 1.3) Manhattan Distance Metric: </a>

## <a href='#1.4'> 1.4) Cosine Distance Metric & Cosine Similarity: </a>

## <a href ='#1.5'> 1.5) Jemsen-Shannon Distance Metric: </a>

# <a href='#I'>  2) CODE IMPLEMENTATION: </a>

## <a href='#2.1'> 2.1) Minkowski  Distance Metric: </a>

## <a href='#2.2'> 2.2) Eucclidean  Distance Metric: </a>

## <a href='#2.3'> 2.3) Mahhatan Distance Metric: <a>
    
## <a href='#2.4'> 2.4) Cosine Distance Metric: </a>
    
## <a href='#2.5'> 2.5) Jemsen-Shannon Distance Metric: </a>
    
    


## <a id='1'> <h1><center> 1. THREOTICAL AND MATHEMATIC BACKGROUND:</center></h1> </a>

## <a id='1.1'> 1.1) Minkowski Distance Metric: </a>

Minkowski Distance is the generalized form of Euclidean and Manhattan Distance.

The formula for Minkowski Distance is given as:

![](https://cdn-images-1.medium.com/max/1067/1*Fb22fnJRABGUANpjcjwEow.png)

The p parameter of the Minkowski Distance metric represents the order of the norm. 

When p is set to 1, the calculation is the same as the Manhattan distance. When p is set to 2, it is the same as the Euclidean distance.

p=1: Manhattan distance.

p=2: Euclidean distance.

p= infinitive, Chebychev Distance

Intermediate values provide a controlled balance between the two measures.

It is common to use Minkowski distance when implementing a machine learning algorithm that uses distance measures as it gives control over the type of distance measure used for real-valued vectors via a hyperparameter “p” that can be tuned.

## <a id='1.2'> 1.2) Euclidean Distance Metric: </a>

Euclidean Distance represents the shortest distance between two points.

Most machine learning algorithms including K-Means use this distance metric to measure the similarity between observations. Let’s say we have two points as shown below:

![](https://cdn-images-1.medium.com/max/1067/1*p1BaA9Px8PimHUuz1V6DMA.png)

So, the Euclidean Distance between these two points A and B will be:

![](https://cdn-images-1.medium.com/max/1067/1*RwxPrdfS0G0w68yAdw-6Cw.png)

Here’s the formula for Euclidean Distance:

![](https://cdn-images-1.medium.com/max/1067/1*_ftwbnr74RtEwnquHPg2rg.png)

We can generalize this for an n-dimensional space as:

![](https://cdn-images-1.medium.com/max/1067/1*lHQbNP1Grabz0ViiPM9uow.png)

Where,

n = number of dimensions
pi, qi = data points

* ## <a id='1.3'> 1.3) Manhattan Distance Metric: </a>

Manhattan Distance is the sum of absolute differences between points across all the dimensions.

We can represent Manhattan Distance as:

![](https://cdn-images-1.medium.com/max/1067/1*kYsOWlz9d7VFWEBYI8Cudg.png)

Since the above representation is 2 dimensional, to calculate Manhattan Distance, we will take the sum of absolute distances in both the x and y directions. So, the Manhattan distance in a 2-dimensional space is given as:

![](https://cdn-images-1.medium.com/max/1067/1*i3kQozqYjq7fj5IHXPlo5g.png)

And the generalized formula for an n-dimensional space is given as:

![](https://cdn-images-1.medium.com/max/1067/1*9A3Ni7-UQ4NjDQd4e1kXaA.png)

Where,

n = number of dimensions
pi, qi = data points

## <a id='1.4'> 1.4) Cosine Distance & Cosine Similarity Metric: </a>

Cosine distance & Cosine Similarity metric is mainly used to find similarities between two data points. As the cosine distance between the data points increases, the cosine similarity, or the amount of similarity decreases, and vice versa.

![](https://miro.medium.com/max/1236/1*23O5Hck8v5flOF9LtvGr8A.png)

In the above figure, imagine the value of θ to be 60 degrees, then by cosine similarity formula, Cos 60 =0.5 and Cosine distance is 1- 0.5 = 0.5. Therefore the points are 50% similar to each other.

## <a id='1.5'> 1.5) Jensen-Shannon Distance Metric: </a>

The Jensen functions is the right metric for calculating distances between probability distribution

![](https://miro.medium.com/max/1400/1*viATYZeg9SiT-ZdzYGjKYA.png)

## <a id='2'> <h1><center> 2. CODE IMPLEMENTATION:</center></h1> </a>

#  <h1><center>Dataset Overview: </h1></center>



In [2]:
 import pandas as pd # Import library
df=pd.read_csv('../input/medical-ratio1/Medical Ratio.csv')
pd.set_option('display.max_columns', None) # display all feature
df=df.drop([38])
df

COMPANY      A01C      A01G      A01H      A01K      A01M  \
0    ASKA PHARMA HOLDINGS  0.000000  0.000000  0.000000  0.000000  0.000756   
1     ASTELLAS PHARMA INC  0.000000  0.000000  0.000000  0.002385  0.000000   
2        BIOFERMIN PHARMA  0.000000  0.000000  0.000000  0.005208  0.000000   
3           CHUGAI PHARMA  0.000194  0.000000  0.000000  0.013030  0.004862   
4   DAIICHI SANKYO CO LTD  0.000000  0.000079  0.000000  0.001182  0.000000   
5      DAINIPPON SUMITOMO  0.000000  0.000000  0.000137  0.001918  0.000000   
6            DAITO PHARMA  0.000000  0.000000  0.000000  0.000000  0.000000   
7   EIKEN CHEMICAL CO LTD  0.000000  0.000000  0.000000  0.000000  0.000000   
8            EISAI CO LTD  0.000000  0.000000  0.000000  0.002617  0.000000   
9             FUJI PHARMA  0.000000  0.000000  0.000000  0.000000  0.000000   
10            FUSO PHARMA  0.000000  0.000000  0.000000  0.012255  0.000000   
11       HISAMITSU PHARMA  0.000000  0.000000  0.000000  0.000000  0.000000   
12             JCR PHARMA  0.000000  0.000000  0.000000  0.005929  0.000000   
13           KAKEN PHARMA  0.001984  0.000992  0.000000  0.000000  0.000992   
14          KISSEI PHARMA  0.000000  0.000000  0.000000  0.000324  0.000000   
15    KYORIN HOLDINGS INC  0.000000  0.000000  0.000000  0.000273  0.000000   
16     KYOWA KIRIN CO LTD  0.000000  0.000556  0.000000  0.007559  0.000000   
17         MOCHIDA PHARMA  0.000000  0.000000  0.000000  0.003459  0.000000   
18       NICHI-IKO PHARMA  0.000000  0.000000  0.000000  0.000000  0.000000   
19       NIPPON CHEMIPHAR  0.000000  0.000000  0.000000  0.000000  0.000000   
20        NIPPON SHINYAKU  0.000000  0.000000  0.000000  0.001577  0.000000   
21          NISSUI PHARMA  0.000000  0.000000  0.000000  0.000000  0.000000   
22             ONO PHARMA  0.000000  0.000000  0.000000  0.001142  0.000000   
23        OTSUKA HOLDINGS  0.000243  0.000243  0.000081  0.002272  0.000000   
24         PEPTIDREAM INC  0.000000  0.000000  0.000000  0.000000  0.000000   
25           ROHTO PHARMA  0.000000  0.000000  0.000000  0.000311  0.000000   
26          SANTEN PHARMA  0.000000  0.000000  0.000000  0.001526  0.000000   
27   SAWAI GROUP HOLDINGS  0.000000  0.000000  0.000000  0.000000  0.000000   
28         SEIKAGAKU CORP  0.000000  0.000000  0.000000  0.005307  0.000000   
29      SHIONOGI & CO LTD  0.000000  0.000000  0.000000  0.002901  0.000000   
30           TAIKO PHARMA  0.000000  0.000000  0.000000  0.000000  0.000000   
31          TAISHO PHARMA  0.000000  0.000000  0.000000  0.000838  0.000000   
32   TAKEDA PHARMA CO LTD  0.000000  0.000000  0.000000  0.002877  0.000000   
33           TORII PHARMA  0.000000  0.000000  0.000000  0.000000  0.000000   
34            TOWA PHARMA  0.000000  0.000000  0.000000  0.000000  0.000000   
35           TSUMURA & CO  0.001988  0.025845  0.000000  0.000000  0.000000   
36           ZERIA PHARMA  0.000000  0.000000  0.000000  0.000000  0.000000   
37                  RATIO  0.000119  0.000749  0.000006  0.002024  0.000179   

        A01N      A01P      A21D      A23B      A23C      A23D      A23F  \
0   0.006803  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
1   0.003634  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
3   0.012252  0.000389  0.000000  0.000000  0.000000  0.000000  0.000000   
4   0.002443  0.000315  0.000000  0.000000  0.000000  0.000000  0.000000   
5   0.001096  0.000137  0.000000  0.000000  0.000000  0.000000  0.000000   
6   0.012048  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
7   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
8   0.005949  0.000119  0.000000  0.000000  0.000000  0.000238  0.000000   
9   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
10  0.000000  0.000000  0.000000  0.012255  0.000000  0.000000  0.000000   
11  0.000255  0.000000  

In [3]:
df.shape

(38, 273)

## The dataset contain 38 observations corresponds to 37 Pharmaceutical Companies and the Ratio Rows here represents the Ratio: **hypothetical proportion (average) firm**  called benchmark company
 

## 273 feature corresponds to 273 sub-IPC (International Patent Classification)

## Each sub-IPC represents different technology 

## Our goal is to define the list of the companies having the same technology to the Ratio. In other words, we calcuate the distance of the sub-IPC of the companies to the Ratio using different Distance Metric. 

## The smaller distance of a company to the Ratio, The higher similarity of that company to the Domain Technology.


In [4]:
df1=df.copy()
df1= df1.drop(columns='COMPANY')
df1

A01C      A01G      A01H      A01K      A01M      A01N      A01P  \
0   0.000000  0.000000  0.000000  0.000000  0.000756  0.006803  0.000000   
1   0.000000  0.000000  0.000000  0.002385  0.000000  0.003634  0.000000   
2   0.000000  0.000000  0.000000  0.005208  0.000000  0.000000  0.000000   
3   0.000194  0.000000  0.000000  0.013030  0.004862  0.012252  0.000389   
4   0.000000  0.000079  0.000000  0.001182  0.000000  0.002443  0.000315   
5   0.000000  0.000000  0.000137  0.001918  0.000000  0.001096  0.000137   
6   0.000000  0.000000  0.000000  0.000000  0.000000  0.012048  0.000000   
7   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
8   0.000000  0.000000  0.000000  0.002617  0.000000  0.005949  0.000119   
9   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
10  0.000000  0.000000  0.000000  0.012255  0.000000  0.000000  0.000000   
11  0.000000  0.000000  0.000000  0.000000  0.000000  0.000255  0.000000   
12  0.000000  0.000000  0.000000  0.005929  0.000000  0.000000  0.000000   
13  0.001984  0.000992  0.000000  0.000000  0.000992  0.027778  0.017857   
14  0.000000  0.000000  0.000000  0.000324  0.000000  0.006155  0.000000   
15  0.000000  0.000000  0.000000  0.000273  0.000000  0.001639  0.000000   
16  0.000000  0.000556  0.000000  0.007559  0.000000  0.003891  0.000111   
17  0.000000  0.000000  0.000000  0.003459  0.000000  0.009608  0.000769   
18  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
19  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
20  0.000000  0.000000  0.000000  0.001577  0.000000  0.002103  0.000000   
21  0.000000  0.000000  0.000000  0.000000  0.000000  0.010204  0.000000   
22  0.000000  0.000000  0.000000  0.001142  0.000000  0.003426  0.000000   
23  0.000243  0.000243  0.000081  0.002272  0.000000  0.008520  0.000974   
24  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
25  0.000000  0.000000  0.000000  0.000311  0.000000  0.001967  0.000725   
26  0.000000  0.000000  0.000000  0.001526  0.000000  0.005015  0.000436   
27  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
28  0.000000  0.000000  0.000000  0.005307  0.000000  0.002274  0.000000   
29  0.000000  0.000000  0.000000  0.002901  0.000000  0.005388  0.000138   
30  0.000000  0.000000  0.000000  0.000000  0.000000  0.054755  0.023055   
31  0.000000  0.000000  0.000000  0.000838  0.000000  0.001303  0.000279   
32  0.000000  0.000000  0.000000  0.002877  0.000000  0.002055  0.000000   
33  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
34  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
35  0.001988  0.025845  0.000000  0.000000  0.000000  0.011928  0.003976   
36  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
37  0.000119  0.000749  0.000006  0.002024  0.000179  0.005473  0.001332   

        A21D      A23B      A23C      A23D      A23F      A23G      A23J  \
0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
1   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
3   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
4   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
5   0.000000  0.000000  0.000000  0.000000  0.000000  0.000411  0.000000   
6   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
7   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
8   0.000000  0.000000  0.000000  0.000238  0.000000  0.000000  0.000000   
9   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
10  0.000000  0.012255  0.000000  0.000000  0.000000  0.000000  0.000000   
11  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
12  0.000000  0.000000  0.000000  0.000000  0.000000  0

## <a id='i=2.1'> 2.1) Minkowski Distance Metric: </a>

In [5]:
from scipy.spatial.distance import cdist
m_dist=pd.DataFrame(cdist(df1,df1,metric='minkowski', p=3))
#diplay the Minkowski Distance:
m_dist

0         1         2         3         4         5         6   \
0   0.000000  0.242261  0.237749  0.180069  0.176319  0.140228  0.358729   
1   0.242261  0.000000  0.062795  0.095677  0.066764  0.103500  0.165448   
2   0.237749  0.062795  0.000000  0.064219  0.078021  0.104922  0.136590   
3   0.180069  0.095677  0.064219  0.000000  0.042147  0.051557  0.186429   
4   0.176319  0.066764  0.078021  0.042147  0.000000  0.036915  0.208291   
5   0.140228  0.103500  0.104922  0.051557  0.036915  0.000000  0.234903   
6   0.358729  0.165448  0.136590  0.186429  0.208291  0.234903  0.000000   
7   0.659834  0.609663  0.564931  0.569528  0.602779  0.604176  0.523980   
8   0.231757  0.032979  0.037338  0.066204  0.059370  0.092339  0.153446   
9   0.544384  0.319174  0.311070  0.367467  0.375688  0.408201  0.234120   
10  0.275366  0.188701  0.141354  0.135744  0.166781  0.173073  0.155177   
11  0.497210  0.261718  0.264053  0.321822  0.322981  0.357650  0.214058   
12  0.329338  0.113537  0.102303  0.153144  0.162860  0.193817  0.107288   
13  0.139077  0.134926  0.105761  0.047412  0.075424  0.048020  0.222879   
14  0.143893  0.098528  0.109163  0.057862  0.034692  0.015174  0.239959   
15  0.049505  0.193022  0.192764  0.136251  0.127468  0.091955  0.318117   
16  0.179225  0.102269  0.061921  0.032312  0.050895  0.054435  0.184415   
17  0.211596  0.052752  0.042755  0.046272  0.044289  0.073654  0.167519   
18  0.416380  0.174343  0.207310  0.258558  0.240934  0.277654  0.208670   
19  0.151159  0.093076  0.109924  0.060944  0.033633  0.025453  0.240697   
20  0.121639  0.132295  0.116316  0.060443  0.067481  0.035559  0.241367   
21  0.133577  0.283329  0.234682  0.188740  0.225390  0.194950  0.291940   
22  0.011631  0.232315  0.229901  0.172571  0.166691  0.130908  0.352443   
23  0.379545  0.182247  0.153385  0.204688  0.227850  0.255272  0.076184   
24  0.617937  0.629701  0.609518  0.590867  0.611688  0.608093  0.605637   
25  0.447212  0.209036  0.218856  0.276001  0.271727  0.307222  0.194648   
26  0.286508  0.063396  0.063522  0.113245  0.115401  0.148118  0.117780   
27  0.483895  0.241891  0.271938  0.324851  0.308504  0.345220  0.262228   
28  0.219782  0.204829  0.161665  0.134458  0.162853  0.150797  0.186353   
29  0.230639  0.020408  0.049286  0.075839  0.054662  0.090513  0.163213   
30  0.387420  0.404455  0.350671  0.327583  0.367031  0.353814  0.334175   
31  0.259279  0.031521  0.056499  0.100289  0.083875  0.119327  0.142363   
32  0.276530  0.049277  0.062488  0.107154  0.102955  0.136987  0.126609   
33  0.150436  0.351517  0.315411  0.261574  0.287957  0.253917  0.393535   
34  0.509625  0.267400  0.293769  0.347873  0.333789  0.370436  0.276397   
35  0.377872  0.171001  0.146104  0.199671  0.219677  0.248321  0.087004   
36  0.385934  0.145280  0.165616  0.219994  0.209743  0.245839  0.163954   
37  0.250223  0.079958  0.037494  0.071584  0.094540  0.121808  0.121114   

          7         8         9         10        11        12        13  \
0   0.659834  0.231757  0.544384  0.275366  0.497210  0.329338  0.139077   
1   0.609663  0.032979  0.319174  0.188701  0.261718  0.113537  0.134926   
2   0.564931  0.037338  0.311070  0.141354  0.264053  0.102303  0.105761   
3   0.569528  0.066204  0.367467  0.135744  0.321822  0.153144  0.047412   
4   0.602779  0.059370  0.375688  0.166781  0.322981  0.162860  0.075424   
5   0.604176  0.092339  0.408201  0.173073  0.357650  0.193817  0.048020   
6   0.523980  0.153446  0.234120  0.155177  0.214058  0.107288  0.222879   
7   0.000000  0.586935  0.655966  0.449094  0.665206  0.574946  0.578963   
8   0.586935  0.000000  0.318274  0.161602  0.265592  0.107571  0.109224   
9   0.655966  0.318274  0.000000  0.350968  0.121787  0.226309  0.410505   
10  0.449094  0.161602  0.350968  0.000000  0.332784  0.188301  0.156204   
11  0.665206  0.265592  0.121787  0.332784  0.000000  0.173609  0.366531   
12  0.574946  0.107571  0.226309  0.188301  0.173609  0.0

In [6]:
# Distance of Compaines to Ratio
np.warnings.filterwarnings('ignore')
Minkowski= m_dist.iloc[37:38, :] 
Minkowski.columns=df['COMPANY']
pd.set_option('display.max_columns', None)
Minkowski.rename(index={37:'Minkowski'},inplace =True)
Minkowski

COMPANY    ASKA PHARMA HOLDINGS  ASTELLAS PHARMA INC  BIOFERMIN PHARMA  \
Minkowski              0.250223             0.079958          0.037494   

COMPANY    CHUGAI PHARMA  DAIICHI SANKYO CO LTD  DAINIPPON SUMITOMO  \
Minkowski       0.071584                0.09454            0.121808   

COMPANY    DAITO PHARMA  EIKEN CHEMICAL CO LTD  EISAI CO LTD  FUJI PHARMA  \
Minkowski      0.121114               0.540648      0.050683     0.303186   

COMPANY    FUSO PHARMA  HISAMITSU PHARMA  JCR PHARMA  KAKEN PHARMA  \
Minkowski     0.119912          0.262577    0.097403      0.112497   

COMPANY    KISSEI PHARMA  KYORIN HOLDINGS INC  KYOWA KIRIN CO LTD  \
Minkowski       0.127284             0.207362            0.073017   

COMPANY    MOCHIDA PHARMA  NICHI-IKO PHARMA  NIPPON CHEMIPHAR  \
Minkowski         0.05413          0.220197          0.127816   

COMPANY    NIPPON SHINYAKU  NISSUI PHARMA  ONO PHARMA  OTSUKA HOLDINGS  \
Minkowski         0.130302       0.222786    0.243158         0.133856   

COMPANY    PEPTIDREAM INC  ROHTO PHARMA  SANTEN PHARMA  SAWAI GROUP HOLDINGS  \
Minkowski        0.588157      0.223268       0.071767              0.282582   

COMPANY    SEIKAGAKU CORP  SHIONOGI & CO LTD  TAIKO PHARMA  TAISHO PHARMA  \
Minkowski        0.145852           0.063787       0.32856        0.07622   

COMPANY    TAKEDA PHARMA CO LTD  TORII PHARMA  TOWA PHARMA  TSUMURA & CO  \
Minkowski              0.074049      0.311139     0.302607      0.129671   

COMPANY    ZERIA PHARMA  RATIO  
Minkowski      0.176409    0.0

In [7]:
# Top 10 companies with Highest Technology Similarity to the Domain Technology:
Minkowski_report = Minkowski.transpose()
Minkowski_report.reset_index()
Minkowski_report = Minkowski_report.sort_values(by='Minkowski',ascending= True).head(11)
Minkowski_report

Minkowski
COMPANY                        
RATIO                  0.000000
BIOFERMIN PHARMA       0.037494
EISAI CO LTD           0.050683
MOCHIDA PHARMA         0.054130
SHIONOGI & CO LTD      0.063787
CHUGAI PHARMA          0.071584
SANTEN PHARMA          0.071767
KYOWA KIRIN CO LTD     0.073017
TAKEDA PHARMA CO LTD   0.074049
TAISHO PHARMA          0.076220
ASTELLAS PHARMA INC    0.079958

## <a href='#2.2'> 2.2) Eucclidean Distance Metric: </a>


In [8]:
from scipy.spatial.distance import cdist
e_dist=pd.DataFrame(cdist(df1,df1,metric='euclidean'))
#diplay the Minkowski Distance:
e_dist

0         1         2         3         4         5         6   \
0   0.000000  0.272009  0.263204  0.196011  0.197795  0.157207  0.387201   
1   0.272009  0.000000  0.077039  0.100655  0.075358  0.116606  0.197775   
2   0.263204  0.077039  0.000000  0.083086  0.087988  0.116699  0.161045   
3   0.196011  0.100655  0.083086  0.000000  0.049979  0.056884  0.210960   
4   0.197795  0.075358  0.087988  0.049979  0.000000  0.042659  0.229105   
5   0.157207  0.116606  0.116699  0.056884  0.042659  0.000000  0.254695   
6   0.387201  0.197775  0.161045  0.210960  0.229105  0.254695  0.000000   
7   0.798273  0.763303  0.716151  0.721185  0.758978  0.758700  0.666547   
8   0.258023  0.036376  0.051379  0.075026  0.064243  0.102575  0.179380   
9   0.607379  0.354045  0.359624  0.419663  0.419412  0.456614  0.292588   
10  0.295855  0.219845  0.174432  0.169598  0.206274  0.209349  0.209170   
11  0.553585  0.287343  0.302144  0.363093  0.358144  0.397711  0.251595   
12  0.366290  0.124880  0.129348  0.178636  0.180319  0.215896  0.149138   
13  0.150748  0.148695  0.130827  0.061654  0.083370  0.058731  0.251512   
14  0.161728  0.110996  0.117823  0.064992  0.038943  0.019358  0.257767   
15  0.056303  0.216623  0.211812  0.145138  0.142506  0.102535  0.341350   
16  0.195140  0.109603  0.074207  0.038931  0.061017  0.062855  0.204923   
17  0.235497  0.057869  0.055955  0.055577  0.049237  0.081996  0.192103   
18  0.467714  0.196588  0.230693  0.287282  0.270883  0.312159  0.228142   
19  0.169782  0.103907  0.118496  0.070555  0.035088  0.027602  0.259402   
20  0.134989  0.148726  0.130461  0.071294  0.076190  0.042204  0.261687   
21  0.166040  0.303602  0.262851  0.208895  0.240725  0.209503  0.344477   
22  0.014038  0.260604  0.253526  0.186299  0.186484  0.146131  0.379216   
23  0.398307  0.200208  0.163716  0.213704  0.235854  0.262381  0.104958   
24  0.697964  0.707976  0.675685  0.656641  0.688442  0.684143  0.666877   
25  0.501209  0.234196  0.249895  0.313970  0.305057  0.345501  0.223256   
26  0.319483  0.069300  0.083417  0.131799  0.127958  0.165637  0.157028   
27  0.543331  0.271867  0.303127  0.361838  0.346341  0.387657  0.281182   
28  0.242277  0.227588  0.190063  0.158115  0.192359  0.182746  0.240257   
29  0.258318  0.021895  0.064361  0.081397  0.061483  0.102396  0.191164   
30  0.457852  0.464682  0.414393  0.398610  0.440348  0.431837  0.383233   
31  0.291244  0.037167  0.061994  0.113489  0.096316  0.135935  0.172846   
32  0.308550  0.054282  0.080018  0.122994  0.114428  0.153433  0.163409   
33  0.170034  0.391770  0.363859  0.300876  0.321019  0.283417  0.457278   
34  0.572176  0.300645  0.329495  0.389074  0.374967  0.416246  0.297589   
35  0.407001  0.182769  0.162236  0.218007  0.229353  0.262227  0.119826   
36  0.432803  0.163574  0.186160  0.248396  0.235994  0.276634  0.184230   
37  0.269106  0.091445  0.054280  0.077706  0.100020  0.125842  0.149045   

          7         8         9         10        11        12        13  \
0   0.798273  0.258023  0.607379  0.295855  0.553585  0.366290  0.150748   
1   0.763303  0.036376  0.354045  0.219845  0.287343  0.124880  0.148695   
2   0.716151  0.051379  0.359624  0.174432  0.302144  0.129348  0.130827   
3   0.721185  0.075026  0.419663  0.169598  0.363093  0.178636  0.061654   
4   0.758978  0.064243  0.419412  0.206274  0.358144  0.180319  0.083370   
5   0.758700  0.102575  0.456614  0.209349  0.397711  0.215896  0.058731   
6   0.666547  0.179380  0.292588  0.209170  0.251595  0.149138  0.251512   
7   0.000000  0.739722  0.764643  0.580600  0.773745  0.714149  0.730534   
8   0.739722  0.000000  0.360213  0.192127  0.296956  0.123991  0.125800   
9   0.764643  0.360213  0.000000  0.415637  0.142149  0.271730  0.468665   
10  0.580600  0.192127  0.415637  0.000000  0.373285  0.210766  0.187245   
11  0.773745  0.296956  0.142149  0.373285  0.000000  0.201170  0.413158   
12  0.714149  0.123991  0.271730  0.210766  0.201170  0.0

In [9]:
# Distance of Compaines to Ratio
np.warnings.filterwarnings('ignore')
Euclidean= e_dist.iloc[37:38, :] 
Euclidean.columns=df['COMPANY']
pd.set_option('display.max_columns', None)
Euclidean.rename(index={37:'Euclidean'},inplace =True)
Euclidean

COMPANY    ASKA PHARMA HOLDINGS  ASTELLAS PHARMA INC  BIOFERMIN PHARMA  \
Euclidean              0.269106             0.091445           0.05428   

COMPANY    CHUGAI PHARMA  DAIICHI SANKYO CO LTD  DAINIPPON SUMITOMO  \
Euclidean       0.077706                0.10002            0.125842   

COMPANY    DAITO PHARMA  EIKEN CHEMICAL CO LTD  EISAI CO LTD  FUJI PHARMA  \
Euclidean      0.149045               0.685486      0.061873     0.352849   

COMPANY    FUSO PHARMA  HISAMITSU PHARMA  JCR PHARMA  KAKEN PHARMA  \
Euclidean     0.144592          0.297575    0.118149       0.12815   

COMPANY    KISSEI PHARMA  KYORIN HOLDINGS INC  KYOWA KIRIN CO LTD  \
Euclidean       0.131142             0.219003            0.084523   

COMPANY    MOCHIDA PHARMA  NICHI-IKO PHARMA  NIPPON CHEMIPHAR  \
Euclidean        0.064811          0.236186          0.132432   

COMPANY    NIPPON SHINYAKU  NISSUI PHARMA  ONO PHARMA  OTSUKA HOLDINGS  \
Euclidean         0.140185       0.251465    0.259803         0.140986   

COMPANY    PEPTIDREAM INC  ROHTO PHARMA  SANTEN PHARMA  SAWAI GROUP HOLDINGS  \
Euclidean        0.647546      0.252672       0.078661              0.307613   

COMPANY    SEIKAGAKU CORP  SHIONOGI & CO LTD  TAIKO PHARMA  TAISHO PHARMA  \
Euclidean        0.170707           0.075617      0.384079       0.085058   

COMPANY    TAKEDA PHARMA CO LTD  TORII PHARMA  TOWA PHARMA  TSUMURA & CO  \
Euclidean              0.081015       0.35863      0.33259      0.148475   

COMPANY    ZERIA PHARMA  RATIO  
Euclidean      0.193025    0.0

In [10]:
# Top 10 companies with Highest Technology Similarity to the Domain Technology:
Euclidean_report = Euclidean.transpose()
Euclidean_report.reset_index()
Euclidean_report= Euclidean_report.sort_values(by='Euclidean',ascending= True).head(11)
Euclidean_report

Euclidean
COMPANY                        
RATIO                  0.000000
BIOFERMIN PHARMA       0.054280
EISAI CO LTD           0.061873
MOCHIDA PHARMA         0.064811
SHIONOGI & CO LTD      0.075617
CHUGAI PHARMA          0.077706
SANTEN PHARMA          0.078661
TAKEDA PHARMA CO LTD   0.081015
KYOWA KIRIN CO LTD     0.084523
TAISHO PHARMA          0.085058
ASTELLAS PHARMA INC    0.091445

## <a id='2.3'> 2.3) Mahattan Distance Metric: </a>

In [11]:
from scipy.spatial.distance import cdist
m_dist=pd.DataFrame(cdist(df1,df1,metric='cityblock'))
#diplay the Minkowski Distance:
m_dist

0         1         2         3         4         5         6   \
0   0.000000  0.423904  0.481529  0.363659  0.327167  0.272253  0.732253   
1   0.423904  0.000000  0.203727  0.206051  0.143851  0.212807  0.495803   
2   0.481529  0.203727  0.000000  0.271988  0.208381  0.271610  0.433107   
3   0.363659  0.206051  0.271988  0.000000  0.147830  0.160751  0.523593   
4   0.327167  0.143851  0.208381  0.147830  0.000000  0.102753  0.492471   
5   0.272253  0.212807  0.271610  0.160751  0.102753  0.000000  0.537838   
6   0.732253  0.495803  0.433107  0.523593  0.492471  0.537838  0.000000   
7   1.818182  1.794332  1.786932  1.773841  1.803683  1.792428  1.818182   
8   0.433029  0.099691  0.189190  0.195440  0.142395  0.206898  0.465810   
9   1.050076  0.663118  0.751603  0.853073  0.784591  0.841507  0.791010   
10  0.570417  0.519755  0.503064  0.472081  0.527131  0.514284  0.695783   
11  0.927736  0.526301  0.643350  0.693065  0.636830  0.698214  0.687852   
12  0.648831  0.247860  0.351779  0.399431  0.364780  0.421799  0.498119   
13  0.291856  0.304389  0.374008  0.212179  0.198062  0.174303  0.610179   
14  0.256992  0.196961  0.247429  0.178545  0.095939  0.066520  0.495346   
15  0.114574  0.340422  0.403893  0.278967  0.230924  0.182973  0.662723   
16  0.361170  0.226968  0.206652  0.150571  0.169426  0.177237  0.497362   
17  0.403406  0.130176  0.175362  0.172667  0.122801  0.182259  0.465286   
18  0.713269  0.338733  0.445056  0.514242  0.444131  0.510443  0.465346   
19  0.266414  0.180222  0.237431  0.186933  0.077336  0.066610  0.500476   
20  0.261169  0.284395  0.281086  0.214186  0.165633  0.118595  0.554090   
21  0.432351  0.597576  0.574830  0.494281  0.502009  0.451440  0.833538   
22  0.043764  0.396211  0.464421  0.337115  0.293413  0.243447  0.717680   
23  0.767552  0.505109  0.458394  0.495748  0.503609  0.519316  0.452937   
24  1.142857  1.142630  1.142857  1.118742  1.136711  1.140391  1.160069   
25  0.811953  0.414626  0.486881  0.592835  0.522598  0.587527  0.561742   
26  0.542611  0.149640  0.268108  0.309059  0.262467  0.317255  0.485963   
27  0.818383  0.437941  0.548217  0.605822  0.534274  0.609530  0.558058   
28  0.448852  0.423176  0.455719  0.390251  0.425420  0.416808  0.697060   
29  0.416975  0.059433  0.194627  0.182813  0.128820  0.203552  0.478668   
30  1.144613  1.124969  1.137668  1.078837  1.132904  1.137142  1.043783   
31  0.481157  0.090847  0.143141  0.272478  0.192181  0.258289  0.446439   
32  0.524344  0.133558  0.259094  0.288537  0.235808  0.299071  0.489583   
33  0.345427  0.721649  0.758929  0.658737  0.615082  0.552933  1.037620   
34  0.865738  0.487253  0.604444  0.657533  0.584879  0.659156  0.602338   
35  0.757084  0.409891  0.455268  0.511493  0.461166  0.533224  0.447915   
36  0.687475  0.314606  0.371499  0.487061  0.408015  0.477972  0.434596   
37  0.503971  0.247517  0.240111  0.213110  0.237997  0.267967  0.465994   

          7         8         9         10        11        12        13  \
0   1.818182  0.433029  1.050076  0.570417  0.927736  0.648831  0.291856   
1   1.794332  0.099691  0.663118  0.519755  0.526301  0.247860  0.304389   
2   1.786932  0.189190  0.751603  0.503064  0.643350  0.351779  0.374008   
3   1.773841  0.195440  0.853073  0.472081  0.693065  0.399431  0.212179   
4   1.803683  0.142395  0.784591  0.527131  0.636830  0.364780  0.198062   
5   1.792428  0.206898  0.841507  0.514284  0.698214  0.421799  0.174303   
6   1.818182  0.465810  0.791010  0.695783  0.687852  0.498119  0.610179   
7   0.000000  1.789389  1.741259  1.552882  1.763164  1.739130  1.804293   
8   1.789389  0.000000  0.715140  0.472827  0.553016  0.297830  0.298462   
9   1.741259  0.715140  0.000000  0.971342  0.388436  0.643281  0.938492   
10  1.552882  0.472827  0.971342  0.000000  0.758338  0.474812  0.511321   
11  1.763164  0.553016  0.388436  0.758338  0.000000  0.438206  0.751699   
12  1.739130  0.297830  0.643281  0.474812  0.438206  0.0

In [12]:
# Distance of Compaines to Ratio
np.warnings.filterwarnings('ignore')
Mahattan= m_dist.iloc[37:38, :] 
Mahattan.columns=df['COMPANY']
pd.set_option('display.max_columns', None)
Mahattan.rename(index={37:'Mahattan'},inplace =True)
Mahattan

COMPANY   ASKA PHARMA HOLDINGS  ASTELLAS PHARMA INC  BIOFERMIN PHARMA  \
Mahattan              0.503971             0.247517          0.240111   

COMPANY   CHUGAI PHARMA  DAIICHI SANKYO CO LTD  DAINIPPON SUMITOMO  \
Mahattan        0.21311               0.237997            0.267967   

COMPANY   DAITO PHARMA  EIKEN CHEMICAL CO LTD  EISAI CO LTD  FUJI PHARMA  \
Mahattan      0.465994               1.722845      0.199519     0.762035   

COMPANY   FUSO PHARMA  HISAMITSU PHARMA  JCR PHARMA  KAKEN PHARMA  \
Mahattan     0.422754           0.59952    0.321094       0.34465   

COMPANY   KISSEI PHARMA  KYORIN HOLDINGS INC  KYOWA KIRIN CO LTD  \
Mahattan        0.26655             0.421888            0.263893   

COMPANY   MOCHIDA PHARMA  NICHI-IKO PHARMA  NIPPON CHEMIPHAR  NIPPON SHINYAKU  \
Mahattan        0.213587          0.462102          0.283941         0.320444   

COMPANY   NISSUI PHARMA  ONO PHARMA  OTSUKA HOLDINGS  PEPTIDREAM INC  \
Mahattan       0.567999    0.480012         0.365514         1.10901   

COMPANY   ROHTO PHARMA  SANTEN PHARMA  SAWAI GROUP HOLDINGS  SEIKAGAKU CORP  \
Mahattan      0.516326       0.220144              0.562041        0.421285   

COMPANY   SHIONOGI & CO LTD  TAIKO PHARMA  TAISHO PHARMA  \
Mahattan           0.221603      1.042537       0.243636   

COMPANY   TAKEDA PHARMA CO LTD  TORII PHARMA  TOWA PHARMA  TSUMURA & CO  \
Mahattan               0.23484      0.775123     0.607597      0.443733   

COMPANY   ZERIA PHARMA  RATIO  
Mahattan      0.407708    0.0

In [13]:
# Top 10 companies with Highest Technology Similarity to the Domain Technology:
Mahattan_report = Mahattan.transpose()
Mahattan_report.reset_index()
Mahattan_report= Mahattan_report.sort_values(by='Mahattan',ascending= True).head(11)
Mahattan_report

Mahattan
COMPANY                        
RATIO                  0.000000
EISAI CO LTD           0.199519
CHUGAI PHARMA          0.213110
MOCHIDA PHARMA         0.213587
SANTEN PHARMA          0.220144
SHIONOGI & CO LTD      0.221603
TAKEDA PHARMA CO LTD   0.234840
DAIICHI SANKYO CO LTD  0.237997
BIOFERMIN PHARMA       0.240111
TAISHO PHARMA          0.243636
ASTELLAS PHARMA INC    0.247517

  ## <a id='2.4'> 2.4) Cosine Distance Metric: </a>

In [14]:
from scipy.spatial.distance import cdist
co_dist=pd.DataFrame(cdist(df1,df1,metric='cosine'))
#diplay the Minkowski Distance:
co_dist

0         1             2         3             4         5   \
0   0.000000  0.073740  6.753885e-02  0.034233  3.832765e-02  0.023539   
1   0.073740  0.000000  3.098855e-03  0.008529  6.011577e-03  0.014504   
2   0.067539  0.003099  1.110223e-16  0.008589  6.302265e-03  0.013270   
3   0.034233  0.008529  8.588671e-03  0.000000  7.232930e-04  0.001653   
4   0.038328  0.006012  6.302265e-03  0.000723  1.110223e-16  0.001977   
5   0.023539  0.014504  1.326975e-02  0.001653  1.977405e-03  0.000000   
6   0.153821  0.028589  2.912372e-02  0.054048  4.955743e-02  0.066280   
7   0.868043  0.847632  8.432179e-01  0.845621  8.539374e-01  0.854690   
8   0.066246  0.000455  2.320453e-03  0.006185  4.101485e-03  0.011320   
9   0.372407  0.132935  1.459391e-01  0.199048  1.888364e-01  0.224060   
10  0.065265  0.030141  2.901233e-02  0.024574  2.619600e-02  0.028337   
11  0.304074  0.085768  9.817208e-02  0.143782  1.347072e-01  0.166366   
12  0.139018  0.014546  2.096866e-02  0.039676  3.541994e-02  0.051963   
13  0.017765  0.022608  2.110535e-02  0.004649  6.279162e-03  0.002314   
14  0.025467  0.012964  1.171638e-02  0.001306  1.459192e-03  0.000222   
15  0.002975  0.047793  4.350814e-02  0.017563  2.024471e-02  0.010016   
16  0.033336  0.010273  6.917637e-03  0.001871  1.751712e-03  0.002098   
17  0.054265  0.002420  3.179069e-03  0.003355  1.985270e-03  0.006886   
18  0.206116  0.035786  4.516572e-02  0.077275  7.007155e-02  0.094166   
19  0.028226  0.011096  1.056391e-02  0.001048  8.624591e-04  0.000337   
20  0.015638  0.023621  1.917628e-02  0.005192  6.155559e-03  0.001705   
21  0.011396  0.101732  9.093698e-02  0.055327  6.203470e-02  0.044565   
22  0.000190  0.067487  6.182840e-02  0.030015  3.378618e-02  0.020017   
23  0.161339  0.022221  2.582077e-02  0.051940  4.753206e-02  0.066170   
24  0.512513  0.555963  5.526671e-01  0.517908  5.327848e-01  0.526744   
25  0.248206  0.056587  6.466164e-02  0.105431  9.703784e-02  0.124737   
26  0.103923  0.003721  8.095826e-03  0.020865  1.757652e-02  0.030225   
27  0.264699  0.063699  7.507491e-02  0.115887  1.072367e-01  0.136330   
28  0.040918  0.046094  4.412208e-02  0.027025  2.995663e-02  0.025459   
29  0.066639  0.000271  2.819669e-03  0.006184  4.096773e-03  0.011470   
30  0.163653  0.226602  2.229677e-01  0.187977  1.961850e-01  0.182470   
31  0.085316  0.001278  2.360358e-03  0.013399  1.012199e-02  0.020226   
32  0.096514  0.002289  6.822805e-03  0.017519  1.417371e-02  0.025944   
33  0.026782  0.168185  1.577460e-01  0.106968  1.145081e-01  0.089301   
34  0.291294  0.077923  9.014066e-02  0.134438  1.252347e-01  0.156367   
35  0.174999  0.027137  3.243624e-02  0.060258  5.420672e-02  0.075060   
36  0.185916  0.027752  3.409717e-02  0.065069  5.835566e-02  0.080393   
37  0.067641  0.001406  2.862019e-03  0.006465  5.112060e-03  0.012395   

          6         7             8             9         10        11  \
0   0.153821  0.868043  6.624604e-02  3.724067e-01  0.065265  0.304074   
1   0.028589  0.847632  4.547253e-04  1.329346e-01  0.030141  0.085768   
2   0.029124  0.843218  2.320453e-03  1.459391e-01  0.029012  0.098172   
3   0.054048  0.845621  6.184604e-03  1.990484e-01  0.024574  0.143782   
4   0.049557  0.853937  4.101485e-03  1.888364e-01  0.026196  0.134707   
5   0.066280  0.854690  1.132033e-02  2.240599e-01  0.028337  0.166366   
6   0.000000  0.861300  3.003166e-02  8.994239e-02  0.073414  0.054923   
7   0.861300  0.000000  8.492956e-01  8.528857e-01  0.699299  0.849071   
8   0.030032  0.849296  1.110223e-16  1.429889e-01  0.027936  0.094036   
9   0.089942  0.852886  1.429889e-01  1.110223e-16  0.203683  0.020879   
10  0.073414  0.699299  2.793637e-02  2.036835e-01  0.000000  0.148222   
11  0.054923  0.849071  9.403580e-02  2.087857e-02  0.148222  0.000000   
12  0.023148  0.832819  1.779452e-02  8.175243e-02  0.049093  0.040746   
13  0.078972  0.860021  1.847974e-02  2.476972e-01  0.032230  0.186809   
14  0.062765 

In [15]:
# Distance of Compaines to Ratio
np.warnings.filterwarnings('ignore')
Cosine= co_dist.iloc[37:38, :] 
Cosine.columns=df['COMPANY']
pd.set_option('display.max_columns', None)
Cosine.rename(index={37:'Cosine'},inplace =True)
Cosine

COMPANY  ASKA PHARMA HOLDINGS  ASTELLAS PHARMA INC  BIOFERMIN PHARMA  \
Cosine               0.067641             0.001406          0.002862   

COMPANY  CHUGAI PHARMA  DAIICHI SANKYO CO LTD  DAINIPPON SUMITOMO  \
Cosine        0.006465               0.005112            0.012395   

COMPANY  DAITO PHARMA  EIKEN CHEMICAL CO LTD  EISAI CO LTD  FUJI PHARMA  \
Cosine       0.029942               0.820049      0.001066     0.142131   

COMPANY  FUSO PHARMA  HISAMITSU PHARMA  JCR PHARMA  KAKEN PHARMA  \
Cosine      0.023402          0.093966    0.017084      0.019469   

COMPANY  KISSEI PHARMA  KYORIN HOLDINGS INC  KYOWA KIRIN CO LTD  \
Cosine        0.011254             0.043167            0.008272   

COMPANY  MOCHIDA PHARMA  NICHI-IKO PHARMA  NIPPON CHEMIPHAR  NIPPON SHINYAKU  \
Cosine         0.002095          0.042601          0.009737         0.020162   

COMPANY  NISSUI PHARMA  ONO PHARMA  OTSUKA HOLDINGS  PEPTIDREAM INC  \
Cosine        0.089246    0.061718         0.023452        0.528265   

COMPANY  ROHTO PHARMA  SANTEN PHARMA  SAWAI GROUP HOLDINGS  SEIKAGAKU CORP  \
Cosine       0.063633       0.005176              0.072344        0.040505   

COMPANY  SHIONOGI & CO LTD  TAIKO PHARMA  TAISHO PHARMA  TAKEDA PHARMA CO LTD  \
Cosine            0.001061      0.209722       0.002753              0.004439   

COMPANY  TORII PHARMA  TOWA PHARMA  TSUMURA & CO  ZERIA PHARMA  RATIO  
Cosine       0.157148     0.087092      0.030821      0.033335    0.0

In [16]:
# Top 10 companies with Highest Technology Similarity to the Domain Technology:
Cosine_report = Cosine.transpose()
Cosine_report.reset_index()
Cosine_report =Cosine_report.sort_values(by='Cosine',ascending= True).head(11)
Cosine_report

Cosine
COMPANY                        
RATIO                  0.000000
SHIONOGI & CO LTD      0.001061
EISAI CO LTD           0.001066
ASTELLAS PHARMA INC    0.001406
MOCHIDA PHARMA         0.002095
TAISHO PHARMA          0.002753
BIOFERMIN PHARMA       0.002862
TAKEDA PHARMA CO LTD   0.004439
DAIICHI SANKYO CO LTD  0.005112
SANTEN PHARMA          0.005176
CHUGAI PHARMA          0.006465

  ## <a id='2.5'> 2.5) Jensenshannon Distance Metric: </a>

In [17]:
from scipy.spatial.distance import cdist
j_dist=pd.DataFrame(cdist(df1,df1,metric='jensenshannon'))
#diplay the Minkowski Distance:
j_dist

0         1         2         3         4         5         6   \
0   0.000000  0.176707  0.240561  0.201697  0.158454  0.145687  0.353941   
1   0.176707  0.000000  0.197953  0.152161  0.099009  0.119886  0.317183   
2   0.240561  0.197953  0.000000  0.242669  0.199823  0.207611  0.318628   
3   0.201697  0.152161  0.242669  0.000000  0.139680  0.160492  0.332334   
4   0.158454  0.099009  0.199823  0.139680  0.000000  0.100718  0.320668   
5   0.145687  0.119886  0.207611  0.160492  0.100718  0.000000  0.328143   
6   0.353941  0.317183  0.318628  0.332334  0.320668  0.328143  0.000000   
7   0.735567  0.714975  0.712645  0.705209  0.722196  0.718126  0.741454   
8   0.198856  0.117892  0.205637  0.158488  0.118018  0.132532  0.304148   
9   0.442472  0.351802  0.371850  0.400177  0.379711  0.389978  0.412386   
10  0.339372  0.293052  0.317960  0.294929  0.306309  0.304519  0.413154   
11  0.374738  0.260488  0.321280  0.305484  0.286570  0.302700  0.381626   
12  0.291492  0.205662  0.271373  0.230090  0.224475  0.233052  0.360654   
13  0.181660  0.190952  0.273568  0.197096  0.180099  0.184192  0.353118   
14  0.123742  0.119165  0.197561  0.174275  0.109517  0.110724  0.309406   
15  0.109079  0.142706  0.229499  0.175589  0.111873  0.117207  0.346686   
16  0.200241  0.175656  0.174179  0.196805  0.168047  0.173533  0.305829   
17  0.191587  0.135247  0.199783  0.166182  0.130903  0.135196  0.314525   
18  0.274980  0.175570  0.245264  0.241623  0.194697  0.219267  0.288485   
19  0.126912  0.115427  0.207828  0.180663  0.105911  0.105061  0.322055   
20  0.171627  0.172629  0.185010  0.191384  0.133996  0.137104  0.302328   
21  0.287836  0.308800  0.304373  0.285722  0.297440  0.294317  0.387639   
22  0.091085  0.156255  0.240516  0.179679  0.126445  0.125168  0.352689   
23  0.367402  0.287006  0.309125  0.289902  0.299435  0.300814  0.353856   
24  0.510817  0.526308  0.539455  0.503893  0.512742  0.515653  0.575975   
25  0.324837  0.216031  0.232996  0.271248  0.231719  0.253037  0.337159   
26  0.243203  0.140761  0.229580  0.187857  0.167498  0.180256  0.332085   
27  0.310401  0.193705  0.264099  0.257640  0.214533  0.244566  0.313781   
28  0.295005  0.280196  0.323688  0.268883  0.275429  0.266258  0.370535   
29  0.185850  0.091788  0.204326  0.143485  0.097904  0.127793  0.306675   
30  0.499842  0.499473  0.528052  0.478928  0.497129  0.497453  0.522473   
31  0.210094  0.136432  0.147340  0.197100  0.135833  0.148064  0.283628   
32  0.233125  0.154086  0.245296  0.185153  0.164958  0.178793  0.329188   
33  0.274400  0.333039  0.360468  0.328063  0.309573  0.300529  0.456891   
34  0.329365  0.212705  0.281494  0.275130  0.235812  0.264298  0.321556   
35  0.350384  0.274275  0.308324  0.297197  0.271589  0.295798  0.355278   
36  0.273078  0.185315  0.203306  0.244740  0.192198  0.213155  0.279393   
37  0.256419  0.186552  0.236534  0.162603  0.177215  0.190907  0.318689   

          7         8         9         10        11        12        13  \
0   0.735567  0.198856  0.442472  0.339372  0.374738  0.291492  0.181660   
1   0.714975  0.117892  0.351802  0.293052  0.260488  0.205662  0.190952   
2   0.712645  0.205637  0.371850  0.317960  0.321280  0.271373  0.273568   
3   0.705209  0.158488  0.400177  0.294929  0.305484  0.230090  0.197096   
4   0.722196  0.118018  0.379711  0.306309  0.286570  0.224475  0.180099   
5   0.718126  0.132532  0.389978  0.304519  0.302700  0.233052  0.184192   
6   0.741454  0.304148  0.412386  0.413154  0.381626  0.360654  0.353118   
7   0.000000  0.715207  0.708381  0.626401  0.705787  0.687299  0.729676   
8   0.715207  0.000000  0.366195  0.278946  0.253786  0.225275  0.197691   
9   0.708381  0.366195  0.000000  0.431346  0.336438  0.367952  0.425903   
10  0.626401  0.278946  0.431346  0.000000  0.321580  0.267590  0.325968   
11  0.705787  0.253786  0.336438  0.321580  0.000000  0.230155  0.322341   
12  0.687299  0.225275  0.367952  0.267590  0.230155  0.0

In [18]:
# Distance of Compaines to Ratio
np.warnings.filterwarnings('ignore')
Jensenshannon= j_dist.iloc[37:38, :] 
Jensenshannon.columns=df['COMPANY']
pd.set_option('display.max_columns', None)
Jensenshannon.rename(index={37:'Jensenshannon'},inplace =True)
Jensenshannon

COMPANY        ASKA PHARMA HOLDINGS  ASTELLAS PHARMA INC  BIOFERMIN PHARMA  \
Jensenshannon              0.256419             0.186552          0.236534   

COMPANY        CHUGAI PHARMA  DAIICHI SANKYO CO LTD  DAINIPPON SUMITOMO  \
Jensenshannon       0.162603               0.177215            0.190907   

COMPANY        DAITO PHARMA  EIKEN CHEMICAL CO LTD  EISAI CO LTD  FUJI PHARMA  \
Jensenshannon      0.318689               0.673251       0.16728     0.370558   

COMPANY        FUSO PHARMA  HISAMITSU PHARMA  JCR PHARMA  KAKEN PHARMA  \
Jensenshannon     0.268057          0.274432    0.224681      0.235602   

COMPANY        KISSEI PHARMA  KYORIN HOLDINGS INC  KYOWA KIRIN CO LTD  \
Jensenshannon        0.21722             0.232232            0.199395   

COMPANY        MOCHIDA PHARMA  NICHI-IKO PHARMA  NIPPON CHEMIPHAR  \
Jensenshannon        0.188121          0.251925          0.224846   

COMPANY        NIPPON SHINYAKU  NISSUI PHARMA  ONO PHARMA  OTSUKA HOLDINGS  \
Jensenshannon         0.208192       0.267951    0.243234         0.231956   

COMPANY        PEPTIDREAM INC  ROHTO PHARMA  SANTEN PHARMA  \
Jensenshannon        0.510268        0.2416        0.17931   

COMPANY        SAWAI GROUP HOLDINGS  SEIKAGAKU CORP  SHIONOGI & CO LTD  \
Jensenshannon              0.268483        0.272537           0.176632   

COMPANY        TAIKO PHARMA  TAISHO PHARMA  TAKEDA PHARMA CO LTD  \
Jensenshannon      0.464366       0.196007              0.202306   

COMPANY        TORII PHARMA  TOWA PHARMA  TSUMURA & CO  ZERIA PHARMA  RATIO  
Jensenshannon      0.348548      0.27393      0.272498      0.228005    0.0

In [19]:
# Top 10 companies with Highest Technology Similarity to the Domain Technology:
Jensenshannon_report = Jensenshannon.transpose()
Jensenshannon_report.reset_index()
Jensenshannon_report =Jensenshannon_report.sort_values(by='Jensenshannon',ascending= True).head(11)
Jensenshannon_report

Jensenshannon
COMPANY                             
RATIO                       0.000000
CHUGAI PHARMA               0.162603
EISAI CO LTD                0.167280
SHIONOGI & CO LTD           0.176632
DAIICHI SANKYO CO LTD       0.177215
SANTEN PHARMA               0.179310
ASTELLAS PHARMA INC         0.186552
MOCHIDA PHARMA              0.188121
DAINIPPON SUMITOMO          0.190907
TAISHO PHARMA               0.196007
KYOWA KIRIN CO LTD          0.199395

In [20]:
from IPython.display import display, HTML

css = """
.output {
    flex-direction: row;
}
"""

HTML('<style>{}</style>'.format(css))

  ## <a id='2.6'> 2.6) : </a> DISTANCE METRIC RESULT SUMMAY:

In [21]:
from IPython.display import display_html 


df1_style = Minkowski_report.style.set_table_attributes("style='display:inline; margin-right:20px;'").set_caption("DF One")
df2_style = Euclidean_report.style.set_table_attributes("style='display:inline'").set_caption("DF Two")
df3_style = Mahattan_report.style.set_table_attributes("style='display:inline'").set_caption("DF Three")
df4_style = Cosine_report.style.set_table_attributes("style='display:inline'").set_caption("DF Four")
df5_style = Jensenshannon_report.style.set_table_attributes("style='display:inline'").set_caption("DF Five")
display_html(df1_style._repr_html_() + df2_style._repr_html_() + df3_style._repr_html_()+ df4_style._repr_html_()+df5_style._repr_html_(), raw=True)

DF One 
 
 
   
 Minkowski 
 
 
 COMPANY 
   
 
 
 
 
 RATIO 
 0.000000 
 
 
 BIOFERMIN PHARMA 
 0.037494 
 
 
 EISAI CO LTD 
 0.050683 
 
 
 MOCHIDA PHARMA 
 0.054130 
 
 
 SHIONOGI & CO LTD 
 0.063787 
 
 
 CHUGAI PHARMA 
 0.071584 
 
 
 SANTEN PHARMA 
 0.071767 
 
 
 KYOWA KIRIN CO LTD 
 0.073017 
 
 
 TAKEDA PHARMA CO LTD 
 0.074049 
 
 
 TAISHO PHARMA 
 0.076220 
 
 
 ASTELLAS PHARMA INC 
 0.079958 
 
 
 

 
 DF Two 
 
 
   
 Euclidean 
 
 
 COMPANY 
   
 
 
 
 
 RATIO 
 0.000000 
 
 
 BIOFERMIN PHARMA 
 0.054280 
 
 
 EISAI CO LTD 
 0.061873 
 
 
 MOCHIDA PHARMA 
 0.064811 
 
 
 SHIONOGI & CO LTD 
 0.075617 
 
 
 CHUGAI PHARMA 
 0.077706 
 
 
 SANTEN PHARMA 
 0.078661 
 
 
 TAKEDA PHARMA CO LTD 
 0.081015 
 
 
 KYOWA KIRIN CO LTD 
 0.084523 
 
 
 TAISHO PHARMA 
 0.085058 
 
 
 ASTELLAS PHARMA INC 
 0.091445 
 
 
 

 
 DF Three 
 
 
   
 Mahattan 
 
 
 COMPANY 
   
 
 
 
 
 RATIO 
 0.000000 
 
 
 EISAI CO LTD 
 0.199519 
 
 
 CHUGAI PHARMA 
 0.213110 
 
 
 MOCHIDA PHARMA 
 0.213587 
 
 
 SANTEN PHARMA 
 0.220144 
 
 
 SHIONOGI & CO LTD 
 0.221603 
 
 
 TAKEDA PHARMA CO LTD 
 0.234840 
 
 
 DAIICHI SANKYO CO LTD 
 0.237997 
 
 
 BIOFERMIN PHARMA 
 0.240111 
 
 
 TAISHO PHARMA 
 0.243636 
 
 
 ASTELLAS PHARMA INC 
 0.247517 
 
 
 

 
 DF Four 
 
 
   
 Cosine 
 
 
 COMPANY 
   
 
 
 
 
 RATIO 
 0.000000 
 
 
 SHIONOGI & CO LTD 
 0.001061 
 
 
 EISAI CO LTD 
 0.001066 
 
 
 ASTELLAS PHARMA INC 
 0.001406 
 
 
 MOCHIDA PHARMA 
 0.002095 
 
 
 TAISHO PHARMA 
 0.002753 
 
 
 BIOFERMIN PHARMA 
 0.002862 
 
 
 TAKEDA PHARMA CO LTD 
 0.004439 
 
 
 DAIICHI SANKYO CO LTD 
 0.005112 
 
 
 SANTEN PHARMA 
 0.005176 
 
 
 CHUGAI PHARMA 
 0.006465 
 
 
 

 
 DF Five 
 
 
   
 Jensenshannon 
 
 
 COMPANY 
   
 
 
 
 
 RATIO 
 0.000000 
 
 
 CHUGAI PHARMA 
 0.162603 
 
 
 EISAI CO LTD 
 0.167280 
 
 
 SHIONOGI & CO LTD 
 0.176632 
 
 
 DAIICHI SANKYO CO LTD 
 0.177215 
 
 
 SANTEN PHARMA 
 0.179310 
 
 
 ASTELLAS PHARMA INC 
 0.186552 
 
 
 MOCHIDA PHARMA 
 0.188121 
 
 
 DAINIPPON SUMITOMO 
 0.190907 
 
 
 TAISHO PHARMA 
 0.196007 
 
 
 KYOWA KIRIN CO LTD 
 0.199395